File: src/main/java/com/mikea/bayes/VarAssignment.java
package com.mikea.bayes;

import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;

import static com.google.common.collect.Lists.newArrayList;

/**
 * @author mike.aizatsky@gmail.com
 */
public class VarAssignment implements Iterable<Var> {
    // TODO: varSet?
    private final Var[] vars;
    private final int[] values;

    public VarAssignment(Var[] vars, int[] values) {
        this.vars = vars;
        this.values = values;
    }

    public int get(Var var) {
        for (int i = 0; i < vars.length; i++) {
            if (vars[i] == var) {
                return values[i];
            }
        }
        throw new NoSuchElementException("Var " + var + " not found in " + this);
    }

    @Override
    public Iterator<Var> iterator() {
        return Arrays.asList(vars).iterator();
    }

    @Override
    public String toString() {
        StringBuilder result = new StringBuilder("{");

        for (int i = 0; i < vars.length; i++) {
            if (i > 0) result.append(", ");
            Var var = vars[i];
            result.append(var.getName());
            result.append("=");
            result.append(var.getValue(values[i]));
        }

        result.append("}");
        return result.toString();
    }

    public boolean containsAll(Var[] otherVars) {
        nextVar: for (Var otherVar : otherVars) {
            for (Var var : vars) {
                if (var == otherVar) continue nextVar;
            }
            return false;
        }
        return true;
    }

    public static Builder at(Var var, int value) {
        return new Builder().at(var, value);
    }

    public static Builder at(Var var, String value) {
        return new Builder().at(var, value);
    }

    public boolean contains(Var var) {
        for (Var v : vars) {
            if (v == var) return true;
        }
        return false;
    }

    public VarAssignment set(Var var, int value) {
        for (int i = 0; i < vars.length; i++) {
            Var v = vars[i];
            if (v == var) {
                int[] newValues = Arrays.copyOf(values, values.length);
                newValues[i] = value;
                return new VarAssignment(vars, newValues);
            }
        }

        Var[] newVars = Arrays.copyOf(vars, vars.length + 1);
        newVars[vars.length] = var;
        int[] newValues = Arrays.copyOf(values, values.length + 1);
        newValues[values.length] = value;
        return new VarAssignment(newVars, newValues);
    }

    public VarAssignment set(Var var, String value) {
        return set(var, var.getValueIndex(value));
    }

    public static class Builder {
        private final List<Var> vars = newArrayList();
        private final List<Integer> values = newArrayList();

        public Builder at(Var var, int value) {
            for (int i = 0; i < vars.size(); i++) {
                Var v =  vars.get(i);
                if (v == var) {
                    values.set(i, value);
                    return this;
                }
            }

            vars.add(var);
            values.add(value);
            return this;
        }

        public Builder at(Var var, String value) {
            return at(var, var.getValueIndex(value));
        }


        public VarAssignment build() {
            Var[] v = new Var[vars.size()];
            int[] va = new int[vars.size()];

            for (int i = 0; i < vars.size(); ++i) {
                v[i] = vars.get(i);
                va[i] = values.get(i);
            }

            return new VarAssignment(v, va);
        }
    }
}

File: src/main/java/com/mikea/bayes/VarAssignment.scala
package com.mikea.bayes


object VarAssignment {
  def at(`var`: Var, value: Int): Builder = new Builder().at(`var`, value)

  def at(`var`: Var, value: String): Builder = new Builder().at(`var`, value)

  class Builder {
    private var vars = List.empty[Var]
    private var values = List.empty[Int]

    def at(`var`: Var, value: Int): Builder = {
      val i = vars.indexOf(`var`)
      if (i < 0) {
        vars = vars :+ `var`
        values = values :+ value
      } else {
        values = values.updated(i, value)
      }

      this
    }

    def at(`var`: Var, value: String): Builder = at(`var`, `var`.getValueIndex(value))

    def build(): VarAssignment = {
      new VarAssignment(vars, values)
    }
  }
}

/**
 * @author mike.aizatsky@gmail.com
 */
class VarAssignment(val vars: Seq[Var], val values: Seq[Int])
  extends Iterable[Var] {

  def get(`var`: Var): Int = {
    for (i <- 0 until vars.length if vars(i) == `var`) {
      return values(i)
    }
    throw new NoSuchElementException("Var " + `var` + " not found in " + this)
  }

  override def iterator: Iterator[Var] = vars.iterator

  override def toString(): String = {
    val result = new StringBuilder("{")
    for (i <- 0 until vars.length) {
      if (i > 0) result.append(", ")
      val `var` = vars(i)
      result.append(`var`.name)
      result.append("=")
      result.append(`var`.getValue(values(i)))
    }
    result.append("}")
    result.toString()
  }

  def containsAll(otherVars: Iterable[Var]): Boolean = {
    for (otherVar <- otherVars) {
      if (!vars.contains(otherVar)) return false
    }
    true
  }

  def contains(`var`: Var): Boolean = {
    vars.contains(`var`)
  }

  def set(`var`: Var, value: Int): VarAssignment = {
    val i = vars.indexOf(`var`)
    if (i < 0) {
      new VarAssignment(vars :+ `var`, values :+ value)
    } else {
      new VarAssignment(vars , values.updated(i, value))
    }
  }

  def set(`var`: Var, value: String): VarAssignment = set(`var`, `var`.getValueIndex(value))
}
